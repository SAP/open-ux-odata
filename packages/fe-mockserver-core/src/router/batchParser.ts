function split(input: string) {
    const LF = '\n';
    const CRLF = '\r\n';
    const a = [];

    let pL = 0;
    let p1 = input.indexOf(CRLF, pL);
    let p2 = input.indexOf(LF, pL);

    while (p1 !== -1 || p2 !== -1) {
        if (p1 !== -1 && p1 <= p2) {
            a.push(input.substring(pL, p1));
            pL = p1 + 2;
        } else {
            a.push(input.substring(pL, p2));
            pL = p2 + 1;
        }
        p1 = input.indexOf(CRLF, pL);
        p2 = input.indexOf(LF, pL);
    }
    if (pL < input.length) {
        a.push(input.substring(pL));
    }
    return a;
}

/**
 *
 */
export class BatchContent {
    type: number;
    stringData: string;
    stringSplit: any;
    pos: number;

    public constructor(data: string) {
        this.type = 0;
        this.stringData = data;
        this.stringSplit = split(data);
        this.pos = 0;
    }

    public lookLine(): string {
        return this.stringSplit[this.pos];
    }

    public readLine(): string {
        return this.stringSplit[this.pos++];
    }

    public inc(): void {
        this.pos++;
    }
}

export class Batch {
    changeSetErrorResponse: any;
    parts: (BatchPart | Batch)[];
    public constructor(public isChangeSet: boolean, public boundary: string) {
        if (this.isChangeSet) {
            this.changeSetErrorResponse = null;
        }
        this.parts = [];
    }
}

export function getBoundary(headerValue: string) {
    const l = headerValue.split(';');

    for (const element of l) {
        const ll = element.split('=');
        if (ll[0].trim() === 'boundary') {
            return ll[1];
        }
    }
    return '';
}

function readHeader(line: string) {
    const colPos = line.indexOf(':');
    if (colPos === -1) {
        throw new Error('Invalid header "content-type" in batch part');
    }

    const s0 = line.substring(0, colPos);
    const s1 = line.substring(colPos + 1);
    return {
        name: s0.toLowerCase(),
        value: s1.trim()
    };
}
export type BatchPart = {
    url: string;
    method: string;
    body: any;
    headers: Record<string, string>;
    contentId?: string;
};

function readAppHttp(batchContent: BatchContent, boundary: string): BatchPart {
    const headers: any = {};
    const payload = [];

    let state = 0; //read url
    const rawUrl = batchContent.readLine();

    state = 1; //read header
    let line = batchContent.lookLine();

    while (line !== null && line !== undefined && line.indexOf(boundary) !== 0) {
        if (state === 1) {
            if (line.length === 0) {
                state = 2; //read body
                batchContent.inc();
            } else {
                const h = readHeader(line);
                headers[h.name] = h.value;
                batchContent.inc();
            }
        } else if (state === 2) {
            payload.push(line);
            batchContent.inc();
        }
        line = batchContent.lookLine();
    }
    if (line === undefined) {
        throw new Error('Invalid boundary while parsing batch request. Expect boundary ' + boundary);
    }

    const [method, url] = rawUrl.split(' ');

    let actualBody: any;
    if (payload.length > 0) {
        try {
            actualBody = JSON.parse(payload[0]);
        } catch (e) {
            actualBody = payload[0];
        }
    }
    return {
        url: `/${url}`,
        method: method,
        headers: headers,
        body: actualBody
    };
}

function extractContent(batchContent: BatchContent, headers: any, boundary: string) {
    let content;
    if (!headers['content-type']) {
        throw new Error('Missing header "content-type" in batch part');
    } else if (headers['content-type'] === 'application/http') {
        content = readAppHttp(batchContent, boundary);
        if (headers['content-id']) {
            content.contentId = headers['content-id'];
        }
    } else if (headers['content-type'].indexOf('multipart/mixed;') > -1) {
        const changeSetBoundary = getBoundary(headers['content-type']);
        content = parseBatch(batchContent, changeSetBoundary, true);
    }
    return content;
}

function parsePart(batchContent: BatchContent, boundary: string) {
    const boundaryNext = boundary;
    const boundaryEnd = boundary + '--';
    const headers: any = {};
    let content;

    let state = 1; //read header
    let line = batchContent.lookLine();

    while (line !== null && line !== undefined && line !== boundaryNext && line !== boundaryEnd) {
        if (state === 1) {
            if (line.length === 0) {
                state = 2; //read body
                batchContent.inc();
            } else {
                const h = readHeader(line);
                headers[h.name] = h.value;
                batchContent.inc();
            }
        } else if (state === 2) {
            content = extractContent(batchContent, headers, boundary);
        }
        line = batchContent.lookLine();
    }
    if (line === undefined) {
        throw new Error('Invalid boundary while parsing batch request. Expect boundary ' + boundary);
    }

    return content;
}

export function parseBatch(content: BatchContent, boundary: string, isChangeset: boolean = false) {
    const boundaryNext = '--' + boundary;
    const boundaryEnd = '--' + boundary + '--';

    const batch = new Batch(isChangeset, boundary);

    let part;

    let line = content.readLine();
    while (line !== null && line !== undefined && line !== boundaryNext) {
        //read lines before first boundary
        line = content.readLine();
    }
    if (line === undefined) {
        throw new Error('Invalid boundary while parsing batch request');
    }
    //line is now read boundary
    line = content.lookLine(); //read line behind

    while (line !== null && line !== undefined && line !== boundaryEnd) {
        part = parsePart(content, boundaryNext);
        if (part) {
            batch.parts.push(part);
        }
        line = content.lookLine(); //now on boundary
        if (line === boundaryNext) {
            line = content.readLine(); //consume boundaryNext
        }
    }

    content.readLine(); //consume boundaryEnd
    line = content.lookLine(); //read line behind
    while (line !== null && line !== undefined && line.length === 0) {
        //read empty lines after first boundary end
        content.readLine();
        line = content.lookLine();
    }

    return batch;
}
