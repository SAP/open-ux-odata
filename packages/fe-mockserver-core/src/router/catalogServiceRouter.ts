import type { ServiceConfigEx } from '../api';
import type { IRouter } from 'router';
import Router from 'router';
import type { IncomingMessage, ServerResponse } from 'http';

/**
 * Create a service mocking the catalog service from the ABAP backend.
 *
 * @param servicesConfig the service configuration.
 * @param version the version of the services
 * @returns a sub router
 */
export function catalogServiceRouter(servicesConfig: ServiceConfigEx[], version = 2): IRouter {
    const router: IRouter = new Router();

    router.get('/ServiceCollection', (_req: IncomingMessage, res: ServerResponse) => {
        res.setHeader('content-type', 'application/json');
        res.write(
            JSON.stringify({
                d: {
                    results: servicesConfig.map((serviceConfig) => {
                        return {
                            ID: serviceConfig._internalName
                        };
                    })
                }
            })
        );
        res.end();
    });

    router.get('/ServiceCollection\\(*', (_req: IncomingMessage, res: ServerResponse) => {
        res.setHeader('content-type', 'application/json');
        res.write(
            JSON.stringify({
                d: {
                    results: servicesConfig.map((serviceConfig) => {
                        return {
                            TechnicalName: serviceConfig._internalName,
                            Version: version
                        };
                    })
                }
            })
        );
        res.end();
    });

    return router;
}
