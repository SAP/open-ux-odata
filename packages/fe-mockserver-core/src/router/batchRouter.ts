import type { ServerResponse } from 'http';
import type { NextFunction, NextHandleFunction } from 'connect';
import ODataRequest from '../request/odataRequest';
import type { DataAccess } from '../data/dataAccess';
import * as http from 'http';
import type { IncomingMessageWithTenant } from './serviceRouter';
import type { Batch, BatchPart } from './batchParser';
import { BatchContent, getBoundary, parseBatch } from './batchParser';

export function isPartChangeSet(part: BatchPart | Batch): part is Batch {
    return (part as Batch).isChangeSet;
}
const NL = '\r\n';

async function handlePart(
    partDefinition: BatchPart,
    dataAccess: DataAccess,
    boundary: string,
    tenantId: string
): Promise<string> {
    const partRequest = new ODataRequest({ ...partDefinition, tenantId: tenantId }, dataAccess);
    await partRequest.handleRequest();
    let batchResponse = '';
    batchResponse += `--${boundary}${NL}`;
    batchResponse += `Content-Type: application/http${NL}`;
    batchResponse += `Content-Transfer-Encoding: binary${NL}`;
    if (partDefinition.contentId) {
        batchResponse += `Content-ID: ${partDefinition.contentId}${NL}`;
    }
    batchResponse += NL;
    const responseData = partRequest.getResponseData();
    batchResponse += `HTTP/1.1 ${partRequest.statusCode} ${http.STATUS_CODES[partRequest.statusCode]}${NL}`;
    for (const headerName in partRequest.responseHeaders) {
        batchResponse += `${headerName}: ${partRequest.responseHeaders[headerName]}${NL}`;
    }
    batchResponse += NL; // End of part header
    if (responseData) {
        batchResponse += responseData;
        batchResponse += NL; // End of body content
    }
    batchResponse += NL;
    return batchResponse;
}

/**
 * Creates a router dedicated to batch request handling.
 * @param dataAccess the current DataAccess object
 * @returns a router function for batch handling
 */
export function batchRouter(dataAccess: DataAccess): NextHandleFunction {
    return async (req: IncomingMessageWithTenant, res: ServerResponse, next: NextFunction) => {
        try {
            const boundary = getBoundary(req.headers['content-type'] as string);
            const body = (req as any).body;
            const batchData = parseBatch(new BatchContent(body), boundary);

            let batchResponse = '';

            for (const part of batchData.parts) {
                if (isPartChangeSet(part)) {
                    batchResponse += `--${batchData.boundary}${NL}`;
                    batchResponse += `Content-Type: multipart/mixed; boundary=${part.boundary}${NL}`;
                    batchResponse += NL;
                    for (const changeSetPart of part.parts) {
                        batchResponse += await handlePart(
                            changeSetPart as BatchPart,
                            dataAccess,
                            part.boundary,
                            req.tenantId!
                        );
                    }
                    batchResponse += `--${part.boundary}--${NL}`;
                } else {
                    batchResponse += await handlePart(part, dataAccess, batchData.boundary, req.tenantId!);
                }
            }
            batchResponse += `--${batchData.boundary}--${NL}`;
            res.statusCode = 200;
            res.setHeader('Content-Type', `multipart/mixed; boundary=${batchData.boundary}`);
            res.setHeader('odata-version', dataAccess.getMetadata().getVersion());
            res.write(batchResponse);
            res.end();
        } catch (e) {
            next(e);
        }
    };
}
