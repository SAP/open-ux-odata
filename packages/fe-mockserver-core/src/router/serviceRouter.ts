import type { NextFunction } from 'express';
import Router from 'router';
import type { IRouter } from 'router';
import { getLogger } from '@ui5/logger';
import type { ServiceConfigEx } from '../api';
import type { ServerResponse } from 'http';
import type { IncomingMessage } from 'connect';
import { raw } from 'body-parser';
import { batchRouter } from './batchRouter';
import ODataRequest from '../request/odataRequest';
import type { DataAccess } from '../data/dataAccess';
import { URL } from 'url';

export type IncomingMessageWithTenant = IncomingMessage & {
    tenantId?: string;
};

/**
 * Checks if a CSRF Token is requested and adds it to the header if so
 *
 * UI5 may user both methods HEAD or GET with the service document url to fetch
 * the token.
 *
 * @param _req IncomingMessage
 * @param res ServerResponse
 */
const addCSRFTokenIfRequested = (_req: IncomingMessage, res: ServerResponse) => {
    if (_req.headers['X-CSRF-Token'.toLowerCase()] === 'Fetch') {
        res.setHeader('X-CSRF-Token', '0504-71383');
    }
};

/**
 * Creates the sub router containing the odata protocol processing.
 *
 * @param service
 * @param dataAccess
 * @returns the sub router specific to that odata query
 */
export async function serviceRouter(service: ServiceConfigEx, dataAccess: DataAccess): Promise<IRouter> {
    const router = new Router();
    const log = getLogger('server:ux-fe-mockserver');

    router.head('/', (_req: IncomingMessage, res: ServerResponse, next) => {
        addCSRFTokenIfRequested(_req, res); //HEAD use case
        next();
    });

    // Deal with the $metadata support
    router.get('/\\$metadata', (_req: IncomingMessage, res: ServerResponse) => {
        res.setHeader('Content-Type', 'application/xml');
        if (service.ETag) {
            res.setHeader('ETag', service.ETag);
        }

        res.write(dataAccess.getMetadata().getEdmx());
        res.end();
    });
    router.post('/\\$metadata/reload', (_req: IncomingMessage, res: ServerResponse) => {
        dataAccess.reloadData();
        res.setHeader('Content-Type', 'application/json');
        res.write(JSON.stringify({ message: 'Reload success' }));
        res.end();
    });
    router.get('/', (_req: IncomingMessage, res: ServerResponse) => {
        const data = `<?xml version="1.0" encoding="utf-8"?>
        <app:service xml:lang="en" xml:base="${service.urlPath}/"
            xmlns:app="http://www.w3.org/2007/app"
            xmlns:atom="http://www.w3.org/2005/Atom"
            xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata"
            xmlns:sap="http://www.sap.com/Protocols/SAPData">
            <app:workspace>
            </app:workspace>
            <atom:link rel="self" href="${service.urlPath}/"/>
            <atom:link rel="latest-version" href="${service.urlPath}/"/>
        </app:service>`;
        res.setHeader('Content-Type', 'application/xml');
        addCSRFTokenIfRequested(_req, res); //GET use case
        res.write(data);
        res.end();
    });
    // Standard processing for the incoming message
    router.use((req: IncomingMessageWithTenant, res, next) => {
        const parser = raw({ type: '*/*' });
        const tenantId = req.originalUrl?.startsWith('/tenant-') ? req.originalUrl?.split('/')[1] : 'tenant-default';
        req.tenantId = tenantId;
        const sapClient = req.originalUrl?.includes('sap-client');
        if (sapClient) {
            const parsedUrl = new URL(`http://dummy${req.originalUrl}`);
            req.tenantId = `tenant-${parsedUrl.searchParams.get('sap-client')}`;
        }
        parser(req, res, function () {
            if (
                (req as any).body === null ||
                (typeof (req as any).body === 'object' && Object.keys((req as any).body).length === 0)
            ) {
                (req as any).body = '{}';
            } else {
                (req as any).body = (req as any).body.toString('utf-8');
            }

            if (
                req.headers['content-type'] &&
                (req as any).body &&
                req.headers['content-type'].indexOf('application/json') !== -1
            ) {
                (req as any).body = JSON.parse((req as any).body);
            }
            next();
        });
    });

    router.use('/\\$batch', batchRouter(dataAccess));

    router.route('/*').all(async (req: IncomingMessageWithTenant, res: ServerResponse, next: NextFunction) => {
        try {
            const oDataRequest = new ODataRequest(
                {
                    url: req.url!,
                    tenantId: req.tenantId!,
                    body: (req as any).body,
                    headers: req.headers,
                    method: req.method!
                },
                dataAccess
            );
            await oDataRequest.handleRequest();
            const responseData = oDataRequest.getResponseData();
            res.statusCode = oDataRequest.statusCode;
            for (const responseHeader in oDataRequest.responseHeaders) {
                res.setHeader(responseHeader, oDataRequest.responseHeaders[responseHeader]);
            }
            if (responseData) {
                res.write(responseData);
            }
            res.end();
        } catch (e) {
            next(e);
        }
    });

    router.use('*', (err: any, _req: IncomingMessage, res: ServerResponse, next: NextFunction) => {
        log.error(err);
        if (res.headersSent) {
            return next(err);
        } else {
            res.statusCode = 500;
            res.write(err.message);
            res.end();
        }
    });

    return router;
}
