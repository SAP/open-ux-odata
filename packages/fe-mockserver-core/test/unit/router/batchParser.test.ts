import { parseBatch, BatchContent } from '../../../src/router/batchParser';

describe('Batch Parser', () => {
    it('can parse correct batch content', () => {
        const ymBatch = `--boundary111
Content-Type:application/http
Content-Transfer-Encoding:binary

GET YOLO
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true
--boundary111--
`;
        const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        expect(myBatch).toMatchSnapshot();
    });
    it('can parse correct batch content with search & search-focus', () => {
        const ymBatch = `--boundary111
Content-Type:application/http
Content-Transfer-Encoding:binary

GET YOLO?$skip=0&$top=10&search=foo&search-focus=bar
sap-cancel-on-close: true
sap-contextid-accept: header
Accept:application/json
Accept-Language:en

--boundary111--
`;
        const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        expect(myBatch).toMatchSnapshot();
    });
    it('can parse correct batch content with changeset', () => {
        const ymBatch = `--batch_id-1653467757956-971
Content-Type: multipart/mixed;boundary=changeset_id-1653467757956-972

--changeset_id-1653467757956-972
Content-Type:application/http
Content-Transfer-Encoding:binary
Content-ID:0.0

POST LineItems(ID=1,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
If-Match:W/"2022-05-25T07:18:11.527Z"
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true

{"myParameter":"23"}
--changeset_id-1653467757956-972
Content-Type:application/http
Content-Transfer-Encoding:binary
Content-ID:1.0

POST LineItems(ID=2,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
If-Match:W/"2022-05-25T07:18:11.527Z"
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true

{"myParameter":"23"}
--changeset_id-1653467757956-972
Content-Type:application/http
Content-Transfer-Encoding:binary
Content-ID:2.0

POST LineItems(ID=3,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
If-Match:W/"2022-05-25T07:18:11.527Z"
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true

{"myParameter":"23"}
--changeset_id-1653467757956-972--
--batch_id-1653467757956-971
Content-Type:application/http
Content-Transfer-Encoding:binary

GET LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%201%20and%20IsActiveEntity%20eq%20true HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true


--batch_id-1653467757956-971
Content-Type:application/http
Content-Transfer-Encoding:binary

GET LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%202%20and%20IsActiveEntity%20eq%20true HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true


--batch_id-1653467757956-971
Content-Type:application/http
Content-Transfer-Encoding:binary

GET LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%203%20and%20IsActiveEntity%20eq%20true HTTP/1.1
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true


--batch_id-1653467757956-971--
Group ID: $auto
`;
        const myBatch = parseBatch(new BatchContent(ymBatch), 'batch_id-1653467757956-971');
        expect(myBatch).toMatchSnapshot();
    });
    it('will fail on incorrect batch parse correct batch content - missing content type', () => {
        const ymBatch = `--boundary111
Content-Transfer-Encoding:binary

GET YOLO
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true
--boundary111--
`;
        try {
            const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
    it('will fail on incorrect batch parse correct batch content - missing content type 2', () => {
        const ymBatch = `--boundary111
Content-Type:application/http
Content-Transfer-Encoding:binary

GET YOLO
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
Content-Type:application/json;charset=UTF-8;IEEE754Compatible=true
--boundary111--
`;
        try {
            const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
    it('will fail on incorrect batch parse correct batch content - incorrect batch ', () => {
        const ymBatch = `--boundary111
Content-Type:application/http
Content-Transfer-Encoding:binary

GET YOLO
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
--boundary1131--
`;
        try {
            const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
    it('will fail on incorrect batch parse correct batch content - incorrect batch boundary ', () => {
        const ymBatch = `--boundary11331
Content-Type:application/http
Content-Transfer-Encoding:binary

GET YOLO
Accept:application/json;odata.metadata=minimal;IEEE754Compatible=true
Accept-Language:en
--boundary1131--
`;
        try {
            const myBatch = parseBatch(new BatchContent(ymBatch), 'boundary111');
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
});
