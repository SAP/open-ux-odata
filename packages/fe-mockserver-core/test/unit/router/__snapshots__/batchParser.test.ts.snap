// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batch Parser can parse correct batch content 1`] = `
Batch {
  "boundary": "boundary111",
  "isChangeSet": false,
  "parts": [
    {
      "body": undefined,
      "headers": {
        "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
        "accept-language": "en",
        "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
      },
      "method": "GET",
      "url": "/YOLO",
    },
  ],
}
`;

exports[`Batch Parser can parse correct batch content with changeset 1`] = `
Batch {
  "boundary": "batch_id-1653467757956-971",
  "isChangeSet": false,
  "parts": [
    Batch {
      "boundary": "changeset_id-1653467757956-972",
      "changeSetErrorResponse": null,
      "isChangeSet": true,
      "parts": [
        {
          "body": {
            "myParameter": "23",
          },
          "contentId": "0.0",
          "headers": {
            "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
            "accept-language": "en",
            "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
            "if-match": "W/"2022-05-25T07:18:11.527Z"",
          },
          "method": "POST",
          "url": "/LineItems(ID=1,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet",
        },
        {
          "body": {
            "myParameter": "23",
          },
          "contentId": "1.0",
          "headers": {
            "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
            "accept-language": "en",
            "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
            "if-match": "W/"2022-05-25T07:18:11.527Z"",
          },
          "method": "POST",
          "url": "/LineItems(ID=2,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet",
        },
        {
          "body": {
            "myParameter": "23",
          },
          "contentId": "2.0",
          "headers": {
            "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
            "accept-language": "en",
            "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
            "if-match": "W/"2022-05-25T07:18:11.527Z"",
          },
          "method": "POST",
          "url": "/LineItems(ID=3,IsActiveEntity=true)/sap.fe.manageitems.TechnicalTestingService.testInvocationGroupingChangeSet",
        },
      ],
    },
    {
      "body": "",
      "headers": {
        "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
        "accept-language": "en",
        "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
      },
      "method": "GET",
      "url": "/LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%201%20and%20IsActiveEntity%20eq%20true",
    },
    {
      "body": "",
      "headers": {
        "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
        "accept-language": "en",
        "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
      },
      "method": "GET",
      "url": "/LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%202%20and%20IsActiveEntity%20eq%20true",
    },
    {
      "body": "",
      "headers": {
        "accept": "application/json;odata.metadata=minimal;IEEE754Compatible=true",
        "accept-language": "en",
        "content-type": "application/json;charset=UTF-8;IEEE754Compatible=true",
      },
      "method": "GET",
      "url": "/LineItems?$select=ID,IsActiveEntity,updatedValue&$filter=ID%20eq%203%20and%20IsActiveEntity%20eq%20true",
    },
  ],
}
`;

exports[`Batch Parser can parse correct batch content with search & search-focus 1`] = `
Batch {
  "boundary": "boundary111",
  "isChangeSet": false,
  "parts": [
    {
      "body": undefined,
      "headers": {
        "accept": "application/json",
        "accept-language": "en",
        "sap-cancel-on-close": "true",
        "sap-contextid-accept": "header",
      },
      "method": "GET",
      "url": "/YOLO?$skip=0&$top=10&search=foo&search-focus=bar",
    },
  ],
}
`;

exports[`Batch Parser will fail on incorrect batch parse correct batch content - incorrect batch  1`] = `[Error: Invalid header "content-type" in batch part]`;

exports[`Batch Parser will fail on incorrect batch parse correct batch content - incorrect batch boundary  1`] = `[Error: Invalid boundary while parsing batch request]`;

exports[`Batch Parser will fail on incorrect batch parse correct batch content - missing content type 1`] = `[Error: Missing header "content-type" in batch part]`;
